File: src/main/cpp/branch_elimination_benchmark.cpp
#include <iostream>
#include <chrono>

// We use 'volatile' to prevent the compiler from optimizing 
// the check away at compile-time (simulating a runtime config flag)
volatile bool SAFETY_CHECK_ENABLED = true;

int runStrategy(int input) {
    if (SAFETY_CHECK_ENABLED) {
        return input + 1;
    } else {
        return input + 2;
    }
}

int main() {
    long long sum = 0;
    int iteration = 2000000000;
    
    // --- MEASUREMENT PHASE ---
    auto start = std::chrono::high_resolution_clock::now();

    for (int i = 0; i < iteration; ++i) {
        sum += runStrategy(i);
    }

    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double, std::micro> elapsed = end - start;

    std::cout << "C++ Time: " << elapsed.count() << " us" << std::endl;
    std::cout << "Check: " << sum << std::endl;
    
    return 0;
}

File: src/main/java/BranchEliminationBenchmark.java
public class BranchEliminationBenchmark {

  // Volatile allows us to change this at runtime,
  // but JIT will still try to optimize the loop "optimistically"
  static volatile boolean SAFETY_CHECK_ENABLED = true;

  public static void main(String[] args) {
    long sum = 0;

    // --- MEASUREMENT PHASE ---
    long start = System.nanoTime();

    // We run a massive loop with the check ENABLED
    int iteration = 2_000_000_000;
    for (int i = 0; i < iteration; i++) {
      sum += runStrategy(i);
    }

    long end = System.nanoTime();
    System.out.printf("Java Time: %.6f us%n", (end - start) / 1_000.0);
    System.out.println("Check: " + sum);
  }

  // The JIT will eventually realize SAFETY_CHECK_ENABLED is constant
  // for long periods and may "Inline" this method and remove the IF entirely.
  public static int runStrategy(int input) {
    if (SAFETY_CHECK_ENABLED) {
      return input + 1; // Hot Path
    } else {
      return input + 2; // Cold Path
    }
  }
}
